### DISCUSSION TITLE
Should contractInterface (& overrides?) be removed from `queryKey`?

### DISCUSSION THREAD

**Discussion by gpxl-dev** [wevm/wagmi#880] [Ideas]
At present queryKeys for contractReads are pretty massive, because they contain the entire interface for the contract.

Since the interface will not affect the result, I'd ideally like to see this removed. I think the same applies to overrides, too, at least for reads?

Finally, it would be nice to be able to pass our own queryKey to override the default - is there any reason this shouldn't be added as an option?

**Comment by gpxl-dev** [comment]
Realised this is not included in the hashed key, so what is there is probably fine.

**Comment by gpxl-dev** [comment]
Nah, coming back to this again. The cache gets very large very quickly because the ABIs are stored in it - I don't see this as being useful.

Check `localStorage.get('wagmi.cache')` for even a small app and I think it'll be large.

**Reply by jxom** [reply]
Yeah we should fix this. We may still need part of the ABI to uniquely identify the key (could probably just serialize the function names), but yeah, the whole thing isn't needed.

**Reply by jxom** [reply]
Overrides could probably go too!

**Reply by niratodev** [reply]
@jxom We have experienced the browser crash issue. _We think_ we have narrowed it down to a specific key in local storage.

```
"queryKey": [{ "entity": "signer" }],
"queryHash": "[{\"entity\":\"signer\"}]"
```

This value in local storage is massive and contains many circular references. For example

```
clientState.queries[9].state.data.provider.provider._jsonRpcConnection.stream._readableState.pipes._parent._readableState.pipes._targetWindow.localStorage["wagmi.cache"]
```

Additionally, `wagmi.cache` contains our entire Next JS hydration values. We noticed all of our translations are in the signer cache.

```
clientState.queries[9].state.data.provider.provider._jsonRpcConnection.stream._readableState.pipes._parent._readableState.pipes._targetWindow.next.router.components["/accounts/[address]"].props.pageProps._nextI18Next.initialI18nStore.en.common
```

A few debugging notes:

1. We are uncertain when the `useSigner` query key is written to local storage. We use a signer in many places throughout the app and it doesn't always show in the cache. As soon as it does, it won't be long before a **Chrome-based** browser tab will crash.

2. We also noticed that the signer key remains the same across wallets. Sometimes we see the previous wallet address in local storage (sometimes it is updated). Would it be better to include the wallet address in the query key?

**Reply by niratodev** [reply]
> We think we have narrowed it down to a specific key in local storage.

One more thing - the browser will crash 99% of the time when that key is in local storage for us.

**Reply by gpxl-dev** [reply]
> > We think we have narrowed it down to a specific key in local storage.
> 
> 
> 
> One more thing - the browser will crash 99% of the time when that key is in local storage for us.

I had problems with this key working with the react query devtools...

**Reply by gpxl-dev** [reply]
> We may still need part of the ABI to uniquely identify the key (could probably just serialize the function names), but yeah, the whole thing isn't needed.

@jxom - I'm not sure you need any of the ABI do you? I had to shift to my own implementation and used:

`[ sender.toLowerCase(), chainId, contractAddress.toLowerCase(), functionName, args ]`

`sender` (i.e. `from`) is the only part of overrides I can imagine having an impact. The ABI is constant unless it's an upgradable proxy, and in that specific case I think it's fair to expect UIs to implement a cachebuster migration, or just to not aggressively cache the value by keeping a low staletime. 

Note that keeping the keys pretty simple, and hierarchical (least specific to most specific) makes them easy to invalidate, e.g. `queryClient.invalidateQueries([ account ])` will invalidate all queries specific to one account, and of course you can go to e.g. contract level or the specific method without having to find ABIs etc.

**Reply by jxom** [reply]
@gpxl-dev – yeah, you are right. We probably don't need it at all. I just remembered `addressOrName` (w/ `functionName`) would already be sufficient enough to uniquely identify the key.

**Reply by jxom** [reply]
@niratodev @gpxl-dev – let me do some digging. We will get this fixed.

**Reply by jxom** [reply]
ICYMI – this is fixed in `0.6.4`.
